e2e/pos_test.go

```diff
@@ -0,0 +1,617 @@
+		config.Premine(senderAddr, defaultBalance)
+		config.SetDevStakingAddresses([]types.Address{senderAddr})
+		config.SetIBFTPoS(true)
+	})
+	srv := srvs[0]
+	client := srv.JSONRPC()
+
+	previousAccountBalance := framework.GetAccountBalance(senderAddr, client, t)
+
+	// Check if the stake is present on the SC
+	actualStakingSCBalance, fetchError := framework.GetStakedAmount(senderAddr, client)
+	if fetchError != nil {
+		t.Fatalf("Unable to fetch staking SC balance, %v", fetchError)
+	}
+
+	assert.Equalf(t,
+		bigDefaultStakedBalance.String(),
+		actualStakingSCBalance.String(),
+		"Staked address balance mismatch before unstake exploit",
+	)
+
+	// Required default values
+	numTransactions := 5
+	signer := crypto.NewEIP155Signer(100)
+	currentNonce := 0
+
+	// TxPool client
+	clt := srv.TxnPoolOperator()
+
+	generateTx := func() *types.Transaction {
+		signedTx, signErr := signer.SignTx(&types.Transaction{
+			Nonce:    uint64(currentNonce),
+			From:     types.ZeroAddress,
+			To:       &stakingContractAddr,
+			GasPrice: bigGasPrice,
+			Gas:      framework.DefaultGasLimit,
+			Value:    big.NewInt(0),
+			V:        big.NewInt(1), // it is necessary to encode in rlp,
+			Input:    framework.MethodSig("unstake"),
+		}, senderKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		currentNonce++
+
+		return signedTx
+	}
+
+	zeroEth := framework.EthToWei(0)
+	for i := 0; i < numTransactions; i++ {
+		var msg *txpoolOp.AddTxnReq
+		unstakeTxn := generateTx()
+
+		msg = &txpoolOp.AddTxnReq{
+			Raw: &any.Any{
+				Value: unstakeTxn.MarshalRLP(),
+			},
+			From: types.ZeroAddress.String(),
+		}
+
+		_, addErr := clt.AddTxn(context.Background(), msg)
+		if addErr != nil {
+			t.Fatalf("Unable to add txn, %v", addErr)
+		}
+	}
+
+	// Set up the blockchain listener to catch the added block event
+	blockNum := waitForBlock(t, srv, 1, 0)
+
+	block, blockErr := client.Eth().GetBlockByNumber(web3.BlockNumber(blockNum), true)
+	if blockErr != nil {
+		t.Fatalf("Unable to fetch block")
+	}
+
+	// Find how much the account paid for all the transactions in this block
+	paidFee := big.NewInt(0).Mul(bigGasPrice, big.NewInt(int64(block.GasUsed)))
+
+	// Check the balances
+	actualAccountBalance := framework.GetAccountBalance(senderAddr, client, t)
+	actualStakingSCBalance, fetchError = framework.GetStakedAmount(senderAddr, client)
+	if fetchError != nil {
+		t.Fatalf("Unable to fetch staking SC balance, %v", fetchError)
+	}
+
+	// Make sure the balances match up
+
+	// expBalance = previousAccountBalance + stakeRefund - block fees
+	expBalance := big.NewInt(0).Sub(big.NewInt(0).Add(previousAccountBalance, bigDefaultStakedBalance), paidFee)
+
+	assert.Equalf(t,
+		expBalance.String(),
+		actualAccountBalance.String(),
+		"Account balance mismatch after unstake exploit",
+	)
+
+	assert.Equalf(t,
+		zeroEth.String(),
+		actualStakingSCBalance.String(),
+		"Staked address balance mismatch after unstake exploit",
+	)
```

Comment:
`framework.GetStakedAmount(senderAddr, client)` on L318 returns total staked amount of the whole network.
It shouldn't be zero right? Only one validator performed unstaking here, so other 3 should still have their balances staked

Reply:
If we are running this with only 1 validator, let's change the test to have at least 2 of them please, to make the test realistic.

Reply:
You're right 💯 

I've upped the validator count and made the test better for checking:

eb35d49